rust-agent/Cargo.toml

```diff
@@ -0,0 +1,32 @@
+[package]
+name = "kata-agent"
+version = "0.1.0"
+authors = ["Yang Bo <bo@hyper.sh>"]
+edition = "2018"
+
+[dependencies]
+oci = { path = "oci" }
+rustjail = { path = "rustjail" }
+protocols = { path = "protocols" }
+lazy_static = "1.3.0"
+error-chain = "0.12.1"
+grpcio = { path = "grpc-rs" }
```

Comment:
Why not:

```
grpcio = { git = "https://github.com/alipay/grpc-rs.git" } instead of using a submodule?
```

Reply:
Good idea @sameo !

Reply:
This is what we plan to do when we start this project. However, for some reason, we cannot do it this way. I cannot recall exactly why. Let's have some input from @teawater.

Reply:
@sameo @jodh-intel 
Use submodule but not git because I got a lot of fail with git like:
```
    Updating crates.io index
    Updating git repository `https://github.com/alipay/grpc-rs.git`
error: failed to load source for a dependency on `grpcio`
```
I think it is the bug of rust.

Reply:
Looking at the manifest file documentation, it looks like this crate should be specified using `[dependency]` *and* `[patch.crates-io]`?

- https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#testing-a-bugfix

Reply:
@jodh-intel I got same error.  Does it work ok in your part?

Reply:
Looks like we don't need to patch, we can just use @sameo's suggestion, but for two of the manifests:

```diff
diff --git a/Cargo.toml b/Cargo.toml
index 40407f8..956d3b9 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -10,7 +10,7 @@ rustjail = { path = "rustjail" }
 protocols = { path = "protocols" }
 lazy_static = "1.3.0"
 error-chain = "0.12.1"
-grpcio = { path = "grpc-rs" }
+grpcio = { git = "https://github.com/alipay/grpc-rs" }
 protobuf = "2.6.1"
 futures = "0.1.27"
 libc = "0.2.58"
diff --git a/protocols/Cargo.toml b/protocols/Cargo.toml
index caaab5d..ef3b84b 100644
--- a/protocols/Cargo.toml
+++ b/protocols/Cargo.toml
@@ -5,6 +5,6 @@ authors = ["Hui Zhu <teawater@hyper.sh>"]
 edition = "2018"

 [dependencies]
-grpcio = { path = "../grpc-rs" }
+grpcio = { git = "https://github.com/alipay/grpc-rs" }
 protobuf = "2.6.1"
 futures = "0.1.27"
```

That gets me beyond the error you are seeing, but the build fails for me with:

```
/usr/include/stdlib.h:31:10: fatal error: 'stddef.h' file not found
/usr/include/stdlib.h:31:10: fatal error: 'stddef.h' file not found, err: true
thread 'main' panicked at 'Unable to generate grpc bindings: ()', src/libcore/result.rs:1165:5
```

Reply:
@jodh-intel What about we keep use submodules until we find a way to handle the build issue?

Reply:
> Looks like we don't need to patch, we can just use @sameo's suggestion, but for two of the manifests:
> 
> ```diff
> diff --git a/Cargo.toml b/Cargo.toml
> index 40407f8..956d3b9 100644
> --- a/Cargo.toml
> +++ b/Cargo.toml
> @@ -10,7 +10,7 @@ rustjail = { path = "rustjail" }
>  protocols = { path = "protocols" }
>  lazy_static = "1.3.0"
>  error-chain = "0.12.1"
> -grpcio = { path = "grpc-rs" }
> +grpcio = { git = "https://github.com/alipay/grpc-rs" }
>  protobuf = "2.6.1"
>  futures = "0.1.27"
>  libc = "0.2.58"
> diff --git a/protocols/Cargo.toml b/protocols/Cargo.toml
> index caaab5d..ef3b84b 100644
> --- a/protocols/Cargo.toml
> +++ b/protocols/Cargo.toml
> @@ -5,6 +5,6 @@ authors = ["Hui Zhu <teawater@hyper.sh>"]
>  edition = "2018"
> 
>  [dependencies]
> -grpcio = { path = "../grpc-rs" }
> +grpcio = { git = "https://github.com/alipay/grpc-rs" }
>  protobuf = "2.6.1"
>  futures = "0.1.27"
> ```
> 
> That gets me beyond the error you are seeing, but the build fails for me with:
> 
> ```
> /usr/include/stdlib.h:31:10: fatal error: 'stddef.h' file not found
> /usr/include/stdlib.h:31:10: fatal error: 'stddef.h' file not found, err: true
> thread 'main' panicked at 'Unable to generate grpc bindings: ()', src/libcore/result.rs:1165:5
> ```

It looks like musl-dev is missing? What distribution are you using? Several rpm-based distribution has no musl package, we have to build it from scratch. bonic/unstable has musl packages.

Reply:
That box is using Ubuntu 19.04 and has `musl-dev` and `musl-tools` installed.

Reply:
Problem solved! The system had `gcc`, `g++` and `musl-g++`, but was missing the `clang` and `llvm` packages!